### PR TITLE
fix assert_allclose doesnt check shape

### LINKED ISSUE #47449: torch.testing.assert_allclose doesn't check shapes
## 🐛 Bug
Tensors of different shapes can be considered `allclose`:
```python
import torch

# Two tensors of different shapes
A = torch.zeros((1, 1))
B = torch.zeros((1,))

torch.testing.assert_allclose(A, B) # passes
```

## Expected behavior

The numpy equivalent - which `assert_allclose` [was based on](https://github.com/pytorch/pytorch/pull/6200) - fails, as it should:
```python
import numpy as np

A = np.zeros((1, 1))
B = np.zeros((1,))

np.testing.assert_allclose(A, B) # fails
```

## Environment
```
Collecting environment information...
PyTorch version: 1.5.1
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.4 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.18.2

Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.2.89
GPU models and configuration: 
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti

Nvidia driver version: 440.100
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip] numpy==1.18.5
[pip] pytorch-memlab==0.0.4
[pip] torch==1.5.1
[pip] torchvision==0.6.1
[conda] numpy                     1.18.5                   pypi_0    pypi
[conda] pytorch-memlab            0.0.4                    pypi_0    pypi
[conda] torch                     1.5.1                    pypi_0    pypi
[conda] torchvision               0.6.1                    pypi_0    pypi
```

## Additional context
The culprit is these two lines
```python
    if expected.shape != actual.shape:
        expected = expected.expand_as(actual)
```
in `assert_allclose`. This means it's obviously deliberate that shape's ignored, but the divergence from numpy is surprising. What's even more surprising is that it makes `assert_allclose` noncommutative:
```python
A = torch.zeros((1, 1))
B = torch.zeros((1,))

torch.testing.assert_allclose(A, B) # passes
torch.testing.assert_allclose(B, A) # fails
```
All in all I'm expecting a 'working as intended, #wontfix' here, which is fair enough. But I figured I'd open this issue to catch anyone else who's tripped up by this and goes a-Googling. 

cc @mruberry @VitalyFedyunin

### PR DESCRIPTION
**Pull request by walterddr**
fix assert_allclose doesnt check shape

should fix #47449.

### CODE REVIEW

**Review by samestep — APPROVED** [review]
lgtm 👍 is this BC-breaking? also I saw that in the original issue thread, @mruberry suggested that the torch.testing module be considered more broadly while fixing this issue

edit: never mind, I just realized that my question about BC is moot since this only touches test code

**Comment by mruberry** [comment]
> 



> lgtm 👍 is this BC-breaking? also I saw that in the original issue thread, @mruberry suggested that the torch.testing module be considered more broadly while fixing this issue

No worries there, yet.  

> edit: never mind, I just realized that my question about BC is moot since this only touches test code

When torch.testing is documented that can still be BC-breaking.

File: torch/testing/__init__.py [hunk 1]
```diff
@@ -212,7 +212,8 @@ def assert_allclose(actual, expected, rtol=None, atol=None, equal_nan=True, msg=
     if not isinstance(expected, torch.Tensor):
         expected = torch.tensor(expected, dtype=actual.dtype)
     if expected.shape != actual.shape:
-        expected = expected.expand_as(actual)
+        raise AssertionError("expected tensor shape {0} doesn't match with actual tensor "
```
Comment (walterddr) [inline torch/testing/__init__.py:215]: this is in fact causing lots of failure in fbcode. @mruberry should I proceed to provide a warning instead?

Comment (mruberry) [reply torch/testing/__init__.py:215]: Making it a warning sounds OK for now. When we start to document torch.testing (soon) we'll probably want to update those callsites in FBCode.

**Review by malfet — APPROVED** [review]

File: torch/testing/__init__.py (hunk 1 above)
Comment (walterddr) [reply torch/testing/__init__.py:215]: fixed all test errors internally instead (there weren't that many afterall :-) )

**Pull request closed without merging**